### PR TITLE
Always on top

### PR DESCRIPTION
**Pull request by hristoterezov**

### CODE REVIEW

File: react/features/toolbox/functions.web.js [hunk 1]
```diff
@@ -138,6 +138,19 @@ export function getToolbarClassNames(props: Object) {
...
 
 /**
+ * Indicates if a toolbar button is enabled.
+ *
+ * @param {string} name - The name of the setting section as defined in
+ * interface_config.js.
+ * @returns {boolean} - True to indicate that the given toolbar button
+ * is enabled, false - otherwise.
+ */
+export function isButtonEnabled(name) {
```
Comment (virtuacoplenny) [inline react/features/toolbox/functions.web.js]: What did this change of moving the function provide you? I'm not against it but I also don't understand why you moved it.

File: react/features/toolbox/components/ToolbarButton.web.js [hunk 2]
```diff
@@ -214,11 +217,11 @@ class ToolbarButton extends AbstractToolbarButton {
 
             if (!button.unclickable) {
                 if (button.tooltipText) {
-                    UIUtil.setTooltipText(this.button,
+                    setTooltipText(this.button,
                         button.tooltipText,
                         tooltipPosition);
                 } else {
-                    UIUtil.setTooltip(this.button,
+                    setTooltip(this.button,
```
Comment (virtuacoplenny) [inline react/features/toolbox/components/ToolbarButton.web.js]: Why did you move tooltips? Again, I'm not against this but I don't understand what it gives you, especially with reactification of tooltips in review.

File: react/features/toolbox/components/StatelessToolbar.web.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/* @flow */
+
+import React, { Component } from 'react';
+
+/**
+ * Implements a toolbar in React/Web. It is a strip that contains a set of
+ * toolbar items such as buttons. Toolbar is commonly placed inside of a
+ * Toolbox.
+ *
+ * @class Toolbar
```
Comment (virtuacoplenny) [inline react/features/toolbox/components/StatelessToolbar.web.js]: I think these comments need updating, at least redoing the "class Toolbar" part.

File: react/features/toolbox/components/ToolbarButton.web.js [hunk 4]
```diff
@@ -109,17 +106,17 @@ class ToolbarButton extends AbstractToolbarButton {
...
         return (
-            <a
-                { ...attributes }
-                onClick = { this._onClick }
-                ref = { this._createRefToButton }>
-                { this._renderInnerElementsIfRequired() }
+            <StatelessToolbarButton { ...props }>
                 { this._renderPopups(popups) }
-            </a>
+            </StatelessToolbarButton>
```
Comment (virtuacoplenny) [inline react/features/toolbox/components/ToolbarButton.web.js]: I don't fully understand the naming of "stateless." Could you explain to me why the button/toolbar on master is not stateless? When I think "stateful" I think of something with this.state.

File: conference.js [hunk 5]
```diff
@@ -2343,6 +2345,7 @@ export default {
             'device count: ' + videoDeviceCount);
 
         APP.store.dispatch(setVideoAvailable(available));
+        APP.API.notifyVideoAvailabilityChanged(available);
```
Comment (virtuacoplenny) [inline conference.js]: I'm wondering these API calls are better to do in redux/actions/middleware. However, it would make the logic more difficult to detect while conference.js is still in heavy use. Is it tricky to move these calls out of conference and into the react/redux app world?

File: react/features/always-on-top/AlwaysOnTop.js [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        api.on('audioMuteStatusChanged', this._audioMutedListener);
+        api.on('videoMuteStatusChanged', this._videoMutedListener);
+        api.on('audioEnabledStatusChanged', this._audioEnabledListener);
+        api.on('videoEnabledStatusChanged', this._videoEnabledListener);
+
+        Promise.all([
+            api.isAudioMuted(),
+            api.isVideoMuted(),
+            api.isAudioAvailable(),
+            api.isVideoAvailable()
```
Comment (virtuacoplenny) [inline react/features/always-on-top/AlwaysOnTop.js]: I just noticed a naming difference between isXEnabled vs isXAvailable. Do the mean the same thing? What does "enabled" mean?

File: react/features/toolbox/functions.web.js (hunk 1 above)
Comment (virtuacoplenny) [inline react/features/toolbox/functions.web.js]: I'm guessing now you did it for a lighter package?

File: react/features/always-on-top/AlwaysOnTop.js [hunk 7]
```diff
@@ -0,0 +1,295 @@
+import React, { Component } from 'react';
+
+import StatelessToolbar from '../toolbox/components/StatelessToolbar';
+import StatelessToolbarButton
+    from '../toolbox/components/StatelessToolbarButton';
+
+const { api } = window.alwaysOnTop;
```
Comment (virtuacoplenny) [inline react/features/always-on-top/AlwaysOnTop.js]: Where does window.alwaysOnTop get defined?

File: conference.js (hunk 5 above)
Comment (hristoterezov) [inline conference.js]: I suppose it won't be tricky but personally I would prefer to move the API module first. Anyway it will be another task.

File: react/features/always-on-top/AlwaysOnTop.js (hunk 7 above)
Comment (hristoterezov) [inline react/features/always-on-top/AlwaysOnTop.js]: jitsi-meet-electron-utils project defines it.

File: react/features/always-on-top/AlwaysOnTop.js (hunk 6 above)
Comment (hristoterezov) [inline react/features/always-on-top/AlwaysOnTop.js]: They are absolutely the same. Initially I was using enabled for everything because this was the naming in jitsi-meet. But after that when I was rebasing the branch everything was changed to available. Probably I forgot to rename the events. Will fix it!

File: react/features/toolbox/components/ToolbarButton.web.js (hunk 4 above)
Comment (hristoterezov) [inline react/features/toolbox/components/ToolbarButton.web.js]: I named it "stateless" because it doesn't use this.state, lifecycle methods and redux.

File: react/features/toolbox/functions.web.js (hunk 1 above)
Comment (hristoterezov) [inline react/features/toolbox/functions.web.js]: I refactored a little bit UIUtils in order to decrease the size of always on top bundle. Specifically `isButtonEnabled` was a good candidate to be moved in the toolbox feature. You think it is better to be defined in UIUtils or somewhere else?

File: react/features/toolbox/components/ToolbarButton.web.js (hunk 2 above)
Comment (hristoterezov) [inline react/features/toolbox/components/ToolbarButton.web.js]: I started this as a part of the effort to reduce the size of the always on top bundle. But at the the problem with UIUtils turned out to be related to an unused import. I don't think I need this change but at the end I liked how the tooltips are separated from all other random util methods.

I think I can easily revert change if needed?

I was not aware of the tooltips reactification but I don't see a problem with this. Are you worried that this changes are going to be in conflict with the reactification PR?

File: modules/API/API.js [hunk 8]
```diff
@@ -58,6 +72,26 @@ function initCommands() {
...
+            callback(audioAvailable);
+            break;
+        case 'is-video-available':
+            callback(videoAvailable);
+            break;
+        default:
+            return false;
+        }
+
+        return true;
```
Comment (virtuacoplenny) [inline modules/API/API.js]: What does the return true/false do exactly? I see it used elsewhere too.

File: react/features/toolbox/components/ToolbarButton.web.js (hunk 4 above)
Comment (virtuacoplenny) [inline react/features/toolbox/components/ToolbarButton.web.js]: Okay, so stateless as in like a react stateless component, one with just a render method?

File: react/features/always-on-top/AlwaysOnTop.js [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        )
+        .catch(console.error);
+
+        window.addEventListener('mousemove', this._mouseMove);
+
+        this._hideToolbarAfterTimeout();
+    }
+
+    /**
+     * Removes the mouse move listener.
```
Comment (virtuacoplenny) [inline react/features/always-on-top/AlwaysOnTop.js]: I think this comment needs updating because more than the mouse move listener is removed.

File: react/features/always-on-top/AlwaysOnTop.js [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+            this.setState({
+                audioMuted,
+                videoMuted,
+                audioAvailable,
+                videoAvailable
+            })
+        )
+        .catch(console.error);
+
+        window.addEventListener('mousemove', this._mouseMove);
```
Comment (virtuacoplenny) [inline react/features/always-on-top/AlwaysOnTop.js]: mousemove will cause a lot of events to get captured. There was no problem encountered?

File: doc/api.md [hunk 11]
```diff
@@ -141,6 +141,20 @@ The `listener` parameter is a Function object with one argument that will be not
 
 The following events are currently supported:
 
+* **audioEnabledStatusChanged** - event notifications about audio enabled status changes. The listener will receive an object with the following structure:
```
Comment (hristoterezov) [inline doc/api.md]: No. I forgot to rename it here. Thanks for pointing it! I'll fix it!

File: doc/api.md [hunk 12]
```diff
@@ -196,6 +210,20 @@ changes. The listener will receive an object with the following structure:
 }
 ```
 
+* **videoEnabledStatusChanged** - event notifications about video enabled status changes. The listener will receive an object with the following structure:
```
Comment (hristoterezov) [inline doc/api.md]: No. I forgot to rename it here. Thanks for pointing it! I'll fix it!

File: modules/API/API.js (hunk 8 above)
Comment (hristoterezov) [inline modules/API/API.js]: true means that the message is processed and false means that it is not processed. If a message is not processed we are caching the message. You can see the logic in Transport.emit.

File: react/features/always-on-top/AlwaysOnTop.js (hunk 10 above)
Comment (hristoterezov) [inline react/features/always-on-top/AlwaysOnTop.js]: I haven't noticed any problems. Aren't we doing the same thing for jitsi-meet's toolbar? Do you have something in mind to improve it?

File: react/features/toolbox/components/ToolbarButton.web.js (hunk 4 above)
Comment (hristoterezov) [inline react/features/toolbox/components/ToolbarButton.web.js]: Yes!
